File: class12/java/BD1201-DmlExemplo.java
/*
 * BD1201-DmlExemplo.java
 * Aula 12 - DML via JDBC: INSERT / UPDATE / DELETE / upsert (Oracle e PostgreSQL)
 * Estudo de caso: BookHub - livraria online
 * Prof. Giovane Barcelos
 *
 * Este exemplo mostra como executar as mesmas operacoes DML do BD1201-DML.sql
 * a partir de uma aplicacao Java, usando JDBC com PreparedStatement. O metodo
 * main() troca de banco apenas alterando a URL/driver de conexao — o restante
 * do codigo (SQL parametrizado) e praticamente identico para Oracle e Postgres,
 * exceto pelo comando de upsert (MERGE x INSERT ... ON CONFLICT).
 *
 * Dependencias (Maven):
 *   Oracle:      com.oracle.database.jdbc:ojdbc11
 *   PostgreSQL:  org.postgresql:postgresql
 */

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;

public class BD1201DmlExemplo {

    // Ajuste a URL/usuario/senha conforme o ambiente local
    private static final String ORACLE_URL  = "jdbc:oracle:thin:@localhost:1521/XEPDB1";
    private static final String POSTGRES_URL = "jdbc:postgresql://localhost:5432/bookhub";
    private static final String USER = "bookhub_user";
    private static final String PASSWORD = "bookhub_pass";

    public static void main(String[] args) throws SQLException {
        System.out.println("=== Executando DML no Oracle ===");
        executarDemonstracao(ORACLE_URL, "oracle");

        System.out.println("\n=== Executando DML no PostgreSQL ===");
        executarDemonstracao(POSTGRES_URL, "postgresql");
    }

    private static void executarDemonstracao(String url, String dialeto) throws SQLException {
        try (Connection conn = DriverManager.getConnection(url, USER, PASSWORD)) {
            conn.setAutoCommit(false);

            int novoClienteId = inserirCliente(conn,
                    "Diego Ponteiro", "diego.ponteiro@email.com", "+55-31-95555-4444", LocalDate.now());
            System.out.println("Cliente inserido com id: " + novoClienteId);

            int linhasAtualizadas = reajustarPrecoLivrosFisicos(conn, 0.10);
            System.out.println("Livros fisicos reajustados: " + linhasAtualizadas);

            int linhasRemovidas = removerLivrosAntigos(conn, 1990);
            System.out.println("Livros antigos removidos: " + linhasRemovidas);

            if ("oracle".equals(dialeto)) {
                upsertEstoqueOracle(conn, 3, 25);
            } else {
                upsertEstoquePostgres(conn, 3, 25);
            }

            conn.commit();
            System.out.println("Transacao confirmada (COMMIT) em " + dialeto);
        } catch (SQLException e) {
            System.err.println("Erro na demonstracao DML (" + dialeto + "): " + e.getMessage());
        }
    }

    /** INSERT parametrizado — identico em Oracle e PostgreSQL. */
    private static int inserirCliente(Connection conn, String nome, String email,
                                       String telefone, LocalDate dataCadastro) throws SQLException {
        String sql = "INSERT INTO Cliente (nome, email, telefone, data_cadastro) VALUES (?, ?, ?, ?)";
        try (PreparedStatement stmt = conn.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {
            stmt.setString(1, nome);
            stmt.setString(2, email);
            stmt.setString(3, telefone);
            stmt.setObject(4, dataCadastro);
            stmt.executeUpdate();
            try (ResultSet rs = stmt.getGeneratedKeys()) {
                return rs.next() ? rs.getInt(1) : -1;
            }
        }
    }

    /** UPDATE parametrizado — identico em Oracle e PostgreSQL. */
    private static int reajustarPrecoLivrosFisicos(Connection conn, double percentual) throws SQLException {
        String sql = "UPDATE Livro SET preco = ROUND(preco * (1 + ?), 2) WHERE tipo = 'F'";
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setDouble(1, percentual);
            return stmt.executeUpdate();
        }
    }

    /** DELETE parametrizado — identico em Oracle e PostgreSQL. */
    private static int removerLivrosAntigos(Connection conn, int anoLimite) throws SQLException {
        String sql = "DELETE FROM Livro WHERE ano_publicacao < ?";
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, anoLimite);
            return stmt.executeUpdate();
        }
    }

    /** Upsert usando MERGE (sintaxe Oracle). */
    private static void upsertEstoqueOracle(Connection conn, int livroId, int quantidade) throws SQLException {
        String sql = "MERGE INTO Estoque tgt " +
                "USING (SELECT ? AS livro_id, ? AS quantidade_nova FROM dual) src " +
                "ON (tgt.livro_id = src.livro_id) " +
                "WHEN MATCHED THEN UPDATE SET tgt.quantidade_disponivel = src.quantidade_nova, " +
                "                             tgt.ultima_atualizacao = SYSDATE " +
                "WHEN NOT MATCHED THEN INSERT (livro_id, quantidade_disponivel, ultima_atualizacao) " +
                "                      VALUES (src.livro_id, src.quantidade_nova, SYSDATE)";
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, livroId);
            stmt.setInt(2, quantidade);
            stmt.executeUpdate();
        }
    }

    /** Upsert usando INSERT ... ON CONFLICT (sintaxe PostgreSQL). */
    private static void upsertEstoquePostgres(Connection conn, int livroId, int quantidade) throws SQLException {
        String sql = "INSERT INTO Estoque (livro_id, quantidade_disponivel, ultima_atualizacao) " +
                "VALUES (?, ?, CURRENT_DATE) " +
                "ON CONFLICT (livro_id) DO UPDATE " +
                "SET quantidade_disponivel = EXCLUDED.quantidade_disponivel, " +
                "    ultima_atualizacao = EXCLUDED.ultima_atualizacao";
        try (PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, livroId);
            stmt.setInt(2, quantidade);
            stmt.executeUpdate();
        }
    }
}

File: class12/python/BD1201-DmlExemplo.py
#!/usr/bin/env python3
"""
BD1201-DmlExemplo.py
Aula 12 - DML via DB-API: INSERT / UPDATE / DELETE / upsert (Oracle e PostgreSQL)
Estudo de caso: BookHub - livraria online
Prof. Giovane Barcelos

Equivalente Python do BD1201-DmlExemplo.java: mesma logica, mesmas operacoes,
usando python-oracledb (Oracle) e psycopg2 (PostgreSQL). O SQL parametrizado e
quase identico entre os dois bancos, exceto pelo comando de upsert
(MERGE x INSERT ... ON CONFLICT) e pelo marcador de parametro (":1" no Oracle
vs "%s" no psycopg2).

Dependencias:
    pip install oracledb psycopg2-binary
"""

from datetime import date

import oracledb
import psycopg2

ORACLE_DSN = "localhost:1521/XEPDB1"
POSTGRES_DSN = dict(host="localhost", port=5432, dbname="bookhub")
USER = "bookhub_user"
PASSWORD = "bookhub_pass"


def inserir_cliente_oracle(conn, nome, email, telefone, data_cadastro):
    """INSERT com RETURNING para recuperar a chave gerada (identity) - Oracle."""
    with conn.cursor() as cur:
        novo_id = cur.var(int)
        cur.execute(
            """
            INSERT INTO Cliente (nome, email, telefone, data_cadastro)
            VALUES (:1, :2, :3, :4)
            RETURNING cliente_id INTO :5
            """,
            [nome, email, telefone, data_cadastro, novo_id],
        )
        return novo_id.getvalue()[0]


def inserir_cliente_postgres(conn, nome, email, telefone, data_cadastro):
    """INSERT com RETURNING para recuperar a chave gerada (identity) - PostgreSQL."""
    with conn.cursor() as cur:
        cur.execute(
            """
            INSERT INTO Cliente (nome, email, telefone, data_cadastro)
            VALUES (%s, %s, %s, %s)
            RETURNING cliente_id
            """,
            (nome, email, telefone, data_cadastro),
        )
        return cur.fetchone()[0]


def reajustar_preco_livros_fisicos(conn, percentual, marcador):
    """UPDATE parametrizado - a unica diferenca entre bancos e o marcador de bind."""
    sql = f"UPDATE Livro SET preco = ROUND(preco * (1 + {marcador}), 2) WHERE tipo = 'F'"
    with conn.cursor() as cur:
        cur.execute(sql, (percentual,) if marcador == "%s" else [percentual])
        return cur.rowcount


def remover_livros_antigos(conn, ano_limite, marcador):
    """DELETE parametrizado - identico em ambos os bancos, exceto o marcador."""
    sql = f"DELETE FROM Livro WHERE ano_publicacao < {marcador}"
    with conn.cursor() as cur:
        cur.execute(sql, (ano_limite,) if marcador == "%s" else [ano_limite])
        return cur.rowcount


def upsert_estoque_oracle(conn, livro_id, quantidade):
    """Upsert via MERGE - sintaxe Oracle."""
    with conn.cursor() as cur:
        cur.execute(
            """
            MERGE INTO Estoque tgt
            USING (SELECT :1 AS livro_id, :2 AS quantidade_nova FROM dual) src
               ON (tgt.livro_id = src.livro_id)
            WHEN MATCHED THEN
                 UPDATE SET tgt.quantidade_disponivel = src.quantidade_nova,
                            tgt.ultima_atualizacao = SYSDATE
            WHEN NOT MATCHED THEN
                 INSERT (livro_id, quantidade_disponivel, ultima_atualizacao)
                 VALUES (src.livro_id, src.quantidade_nova, SYSDATE)
            """,
            [livro_id, quantidade],
        )


def upsert_estoque_postgres(conn, livro_id, quantidade):
    """Upsert via INSERT ... ON CONFLICT - sintaxe PostgreSQL."""
    with conn.cursor() as cur:
        cur.execute(
            """
            INSERT INTO Estoque (livro_id, quantidade_disponivel, ultima_atualizacao)
            VALUES (%s, %s, CURRENT_DATE)
            ON CONFLICT (livro_id) DO UPDATE
            SET quantidade_disponivel = EXCLUDED.quantidade_disponivel,
                ultima_atualizacao = EXCLUDED.ultima_atualizacao
            """,
            (livro_id, quantidade),
        )


def executar_demonstracao_oracle():
    print("=== Executando DML no Oracle ===")
    conn = oracledb.connect(user=USER, password=PASSWORD, dsn=ORACLE_DSN)
    try:
        novo_id = inserir_cliente_oracle(
            conn, "Diego Ponteiro", "diego.ponteiro@email.com", "+55-31-95555-4444", date.today()
        )
        print(f"Cliente inserido com id: {novo_id}")

        linhas = reajustar_preco_livros_fisicos(conn, 0.10, ":1")
        print(f"Livros fisicos reajustados: {linhas}")

        removidos = remover_livros_antigos(conn, 1990, ":1")
        print(f"Livros antigos removidos: {removidos}")

        upsert_estoque_oracle(conn, 3, 25)

        conn.commit()
        print("Transacao confirmada (COMMIT) em oracle")
    except oracledb.Error as exc:
        conn.rollback()
        print(f"Erro na demonstracao DML (oracle): {exc}")
    finally:
        conn.close()


def executar_demonstracao_postgres():
    print("=== Executando DML no PostgreSQL ===")
    conn = psycopg2.connect(user=USER, password=PASSWORD, **POSTGRES_DSN)
    try:
        novo_id = inserir_cliente_postgres(
            conn, "Diego Ponteiro", "diego.ponteiro@email.com", "+55-31-95555-4444", date.today()
        )
        print(f"Cliente inserido com id: {novo_id}")

        linhas = reajustar_preco_livros_fisicos(conn, 0.10, "%s")
        print(f"Livros fisicos reajustados: {linhas}")

        removidos = remover_livros_antigos(conn, 1990, "%s")
        print(f"Livros antigos removidos: {removidos}")

        upsert_estoque_postgres(conn, 3, 25)

        conn.commit()
        print("Transacao confirmada (COMMIT) em postgresql")
    except psycopg2.Error as exc:
        conn.rollback()
        print(f"Erro na demonstracao DML (postgresql): {exc}")
    finally:
        conn.close()


def main():
    executar_demonstracao_oracle()
    print()
    executar_demonstracao_postgres()


if __name__ == "__main__":
    main()
